cmd/build.go

```diff
@@ -16,7 +18,13 @@ func newLocalExecuteCommand(config *settings.Config) *cobra.Command {
 			return nil
 		},
 		RunE: func(cmd *cobra.Command, _ []string) error {
-			return local.Execute(cmd.Flags(), config, args)
+			telemetryClient := create_telemetry.CreateTelemetry(config)
+			defer telemetryClient.Close()
+
+			err := local.Execute(cmd.Flags(), config, args)
+			_ = telemetryClient.Track(telemetry.CreateLocalExecuteEvent(create_telemetry.GetCommandInformation(cmd, true)))
```

Comment:
I believe we could even simplify the `PreRun` by putting the logic around the command context in a function `QueueTelemetryEventInCommandContext`

Reply:
Yes, I finally understood why you had different patterns, at the end of the review.

I don't mind calling `Track` at different times in the lifecycle of the command (`PreRun`, `Run`, `PostRun`). I'm more concerned about calling `CreateTelemetry` just about anywhere, in the context of the lifecycle hook. I'd keep your `Track` calls as they are, just move the `CreateTelemetry` and the `telemetryClient.Close()` all the way at the top, as I suggested.

So you can still handle adding the events to the client at any point in the lifecycle, including inside `Execute`, capturing the error for the telemetry event.

----

As far as factorizing the Context stuff goes, I'd quote Rob Pike's [Go Proverbs](https://go-proverbs.github.io/) and say that [a little copying is better than a little dependency](https://www.youtube.com/watch?v=PAAkCSZUG1c&t=568s) and [clear is better than clever](https://www.youtube.com/watch?v=PAAkCSZUG1c&t=875s). In this case, it may be better to just copy-paste the one-liner that retrieves the telemetry client from the context and leave the caller to do with as they feel (`cmd` owns `telemetry`, not the other way around); abstracting stuff about contexts in `telemetry` creates coupling that could make the code harder to maintain.

That said, there is a cleaner pattern, albeit a little over-engineered for our case, IMO:

### `telemetry/telemetry.go`

```
type contextKey string

const telemetryClientContextKey = "telemetryClientContextKey"

func NewContext(ctx context.Context, client telemetry.Client) context.Context {
	return context.WithValue(ctx, telemetryClientContextKey, client)
}

func FromContext(ctx context.Context) (telemetry.Client, bool) {
	client, ok := ctx.Value(telemetryClientContextKey).(telemetry.Client)
	return client, ok
}
```

Very idiomatic Go, but feels a bit too much to me.

Reply:
We have some commands defined in their own package so I'm going for your pattern since it will allow other packages to use the telemetry